src/ir/module-utils.cpp

```diff
@@ -198,10 +198,92 @@ void setIndices(IndexedHeapTypes& indexedTypes) {
   }
 }
 
+InsertOrderedSet<HeapType> getPublicTypeSet(Module& wasm) {
+  InsertOrderedSet<HeapType> publicTypes;
+
+  auto notePublic = [&](HeapType type) {
+    // TODO: Gracefully handle public types with arbitrary public children here
+    // and leave it to separate code to handle erroring or warning if
+    // optimizations would be inhibited.
+    if (type.isBasic()) {
+      return;
+    }
+    // All the rec group members are public as well.
```

Comment:
In the common case wouldn't that make all types public, if they are all in one big rec group?

Reply:
Yes. Consumers should not put their public function types in the same rec groups as their private, internal types.

Reply:
I see. What happens in nominal mode, though? Don't we emulate nominal using a single big rec group in the output?

Reply:
Yes, in the output, but the internal representation in nominal mode only uses singleton rec groups.

Reply:
Wouldn't this workflow break then: optimize in nominal mode, then write a binary, then read and optimize that binary in normal mode?

Reply:
Yes, that would break. But it won't be a problem once we delete nominal mode (as long as producers update to not use single huge rec groups).